tests/core/engine_adapter/test_base.py

```diff
@@ -704,6 +705,314 @@ def test_merge_pandas(make_mocked_engine_adapter: t.Callable):
     )
 
 
+def test_scd_type_2(make_mocked_engine_adapter: t.Callable):
+    adapter = make_mocked_engine_adapter(EngineAdapter)
+
+    adapter.scd_type_2(
+        target_table="target",
+        source_table=t.cast(
+            exp.Select, parse_one("SELECT id, name, price, updated_at FROM source")
+        ),
+        unique_key=["id"],
+        valid_from_name="test_valid_from",
+        valid_to_name="test_valid_to",
+        updated_at_name="test_updated_at",
+        columns_to_types={
+            "id": exp.DataType.Type.INT,
+            "name": exp.DataType.Type.VARCHAR,
+            "price": exp.DataType.Type.DOUBLE,
+            "test_updated_at": exp.DataType.Type.TIMESTAMP,
+        },
+        execution_time=datetime(2020, 1, 1, 0, 0, 0),
+    )
+
+    assert (
+        adapter.cursor.execute.call_args[0][0]
+        == parse_one(
+            """
+CREATE OR REPLACE TABLE "target" AS
+WITH "source" AS (
+  SELECT DISTINCT ON ("id")
+    "id",
+    "name",
+    "price",
+    "test_updated_at"
+  FROM (
+    SELECT
+      "id",
+      "name",
+      "price",
+      "updated_at"
+    FROM "source"
+  ) AS "raw_source"
+), "static" AS (
+  SELECT
+    "id",
+    "name",
+    "price",
+    "test_updated_at"
+  FROM "target"
+  WHERE
+    NOT "test_valid_to" IS NULL
+), "latest" AS (
+  SELECT
+    "id",
+    "name",
+    "price",
+    "test_updated_at"
+  FROM "target"
+  WHERE
+    "test_valid_to" IS NULL
+), "deleted" AS (
+  SELECT
+    "static"."id",
+    "static"."name",
+    "static"."price",
+    "static"."test_updated_at"
+  FROM "static"
+  LEFT JOIN "latest"
+    USING ("id")
+  WHERE
+    "latest"."test_valid_to" IS NULL
+), "latest_deleted" AS (
+  SELECT
+    "id",
+    MAX("test_valid_to") AS "test_valid_to"
+  FROM "deleted"
```

Comment:
this query means that the join on deleted has to be done an extra time, should we add a rank function on deleted to have a latest deleted flag instead?

Reply:
I considered a rank but wasn't sure how universally it was supported across current and future engines while obviously this would be supported. Really this aggregate should be pretty cheap and the result set should be pretty small since it is only deleted rows. So my preference was to stick with what I knew was universally supported in this case.

Reply:
window functions (rank) is supported in every basically every sql engine, all the ones we support i'm sure

the aggregate is not the problem, it's the delete that has to be done multiple times i believe